src/zenml/services/service_monitor.py

```diff
@@ -0,0 +1,228 @@
+        default_factory=HTTPEndpointHealthMonitorConfig
+    )
+
+    def get_healthcheck_uri(
+        self, endpoint: "BaseServiceEndpoint"
+    ) -> Optional[str]:
+        uri = endpoint.status.uri
+        if not uri:
+            return None
+        return f"{uri}{self.config.healthcheck_uri_path or '/'}"
+
+    def check_endpoint_status(
+        self, endpoint: "BaseServiceEndpoint"
+    ) -> Tuple[ServiceState, str]:
+        """Run a HTTP endpoint API healthcheck
+
+        Returns:
+            The operational state of the external HTTP endpoint and an
+            optional message describing that state (e.g. an error message,
+            if an error is encountered while checking the HTTP endpoint
+            status).
+        """
+        from zenml.services.service_endpoint import ServiceEndpointProtocol
+
+        if endpoint.status.protocol not in [
+            ServiceEndpointProtocol.HTTP,
+            ServiceEndpointProtocol.HTTPS,
+        ]:
+            return ServiceState.ERROR, "endpoint protocol is not HTTP nor HTTPS"
+
+        check_uri = self.get_healthcheck_uri(endpoint)
+        if not check_uri:
+            return ServiceState.ERROR, "no HTTP healthcheck URI available"
+
+        logger.debug("Running HTTP healthcheck for URI: %s", check_uri)
+
+        error = ""
+
+        try:
+            if self.config.use_head_request:
+                r = requests.head(
+                    check_uri,
+                    timeout=self.config.http_timeout,
+                )
+            else:
+                r = requests.get(
+                    check_uri,
+                    timeout=self.config.http_timeout,
+                )
+            if r.status_code == self.config.http_status_code:
+                # the endpoint is healthy
+                return ServiceState.ACTIVE, ""
+            error = f"HTTP endpoint healthcheck returned unexpected status code: {r.status_code}"
+        except requests.ConnectionError as e:
+            error = f"HTTP endpoint healthcheck connection error: {str(e)}"
+        except requests.Timeout as e:
+            error = f"HTTP endpoint healthcheck request timed out: {str(e)}"
+        except requests.RequestException as e:
+            error = (
+                f"unexpected error encountered while running HTTP endpoint "
+                f"healthcheck: {str(e)}"
+            )
+
+        return ServiceState.ERROR, error
+
+
+class TCPEndpointHealthMonitorConfig(ServiceEndpointHealthMonitorConfig):
+    """TCP service endpoint health monitor configuration."""
+
+    ...
+
+
+class TCPEndpointHealthMonitor(BaseServiceEndpointHealthMonitor):
+    """TCP service endpoint health monitor.
+
+    Attributes:
+        config: TCP endpoint health monitor configuration
+    """
+
+    config: TCPEndpointHealthMonitorConfig = Field(
+        default_factory=TCPEndpointHealthMonitorConfig
+    )
+
+    @classmethod
+    def port_is_open(cls, hostname: str, port: int) -> bool:
```

Comment:
Great idea ! I found some other similar utility functions there too (TCP port availability, allocation). I'll point the code to them rather than duplicating everything.